File: testdata/fixtures/src/sample.swift
// sample.swift — comprehensive Swift syntax fixture for parser testing.
// Covers: structs, classes, protocols, extensions, generics, opaque types,
// property wrappers, async/await, actors, result builders, pattern matching,
// keyPaths, error handling, closures, @discardableResult, sendable.

import Foundation

// -------------------------------------------------------------------------- //
// Constants & type aliases
// -------------------------------------------------------------------------- //

let maxRetries = 3
let defaultTimeout: TimeInterval = 30
typealias UUID = Foundation.UUID
typealias Handler<T> = (Result<T, Error>) -> Void

// -------------------------------------------------------------------------- //
// Enums with associated values
// -------------------------------------------------------------------------- //

enum Status: String, CaseIterable, Sendable {
    case pending = "pending"
    case running = "running"
    case done    = "done"
    case failed  = "failed"

    var isTerminal: Bool { self == .done || self == .failed }
}

enum AppError: LocalizedError {
    case notFound(resource: String, id: String)
    case validation(field: String, detail: String)
    case network(statusCode: Int, url: URL)
    case wrapped(Error)

    var errorDescription: String? {
        switch self {
        case .notFound(let r, let id):         return "\(r) with id='\(id)' not found"
        case .validation(let f, let d):        return "\(f): \(d)"
        case .network(let code, let url):      return "HTTP \(code) from \(url)"
        case .wrapped(let e):                  return e.localizedDescription
        }
    }
}

// -------------------------------------------------------------------------- //
// Protocols
// -------------------------------------------------------------------------- //

protocol Entity {
    var id: UUID { get }
    var createdAt: Date { get }
}

protocol Repository<Element>: Actor where Element: Entity {
    associatedtype Element
    func findById(_ id: UUID) async throws -> Element
    func list() async throws -> [Element]
    func save(_ entity: Element) async throws -> Element
    func delete(_ id: UUID) async throws
}

protocol Validatable {
    var validationErrors: [String] { get }
    var isValid: Bool { get }
}

extension Validatable {
    var isValid: Bool { validationErrors.isEmpty }
}

// -------------------------------------------------------------------------- //
// Structs
// -------------------------------------------------------------------------- //

struct Point: Hashable, Sendable, CustomStringConvertible {
    let x: Double
    let y: Double

    static let origin = Point(x: 0, y: 0)

    func distance(to other: Point) -> Double {
        sqrt(pow(x - other.x, 2) + pow(y - other.y, 2))
    }

    static func + (lhs: Point, rhs: Point) -> Point {
        Point(x: lhs.x + rhs.x, y: lhs.y + rhs.y)
    }

    static func * (lhs: Point, rhs: Double) -> Point {
        Point(x: lhs.x * rhs, y: lhs.y * rhs)
    }

    var description: String { "(\(x), \(y))" }
}

struct User: Entity, Validatable, Sendable {
    let id: UUID
    let createdAt: Date
    var name: String
    var email: String
    var role: String

    init(name: String, email: String, role: String = "member") {
        self.id = UUID()
        self.createdAt = Date()
        self.name = name
        self.email = email
        self.role = role
    }

    var validationErrors: [String] {
        var errors: [String] = []
        if name.trimmingCharacters(in: .whitespaces).isEmpty { errors.append("name is required") }
        if !email.contains("@") { errors.append("email '\(email)' is invalid") }
        return errors
    }
}

// -------------------------------------------------------------------------- //
// Property wrappers
// -------------------------------------------------------------------------- //

@propertyWrapper
struct Clamped<T: Comparable> {
    private var value: T
    let range: ClosedRange<T>

    init(wrappedValue: T, _ range: ClosedRange<T>) {
        self.range = range
        self.value = min(max(wrappedValue, range.lowerBound), range.upperBound)
    }

    var wrappedValue: T {
        get { value }
        set { value = min(max(newValue, range.lowerBound), range.upperBound) }
    }
}

@propertyWrapper
struct Trimmed {
    private var value: String = ""

    var wrappedValue: String {
        get { value }
        set { value = newValue.trimmingCharacters(in: .whitespaces) }
    }

    init(wrappedValue: String) { self.wrappedValue = wrappedValue }
}

struct Config {
    @Clamped(0...65535) var port: Int = 8080
    @Trimmed var host: String = "localhost"

    var baseURL: String { "http://\(host):\(port)" }
}

// -------------------------------------------------------------------------- //
// Generics & opaque types
// -------------------------------------------------------------------------- //

func zip<A, B>(_ a: [A], _ b: [B]) -> [(A, B)] {
    zip(a, b).map { ($0, $1) }
}

func reduce<T, U>(_ sequence: [T], initial: U, combine: (U, T) -> U) -> U {
    sequence.reduce(initial, combine)
}

func makeAdder(_ n: Int) -> some Numeric & Sendable { n }

// -------------------------------------------------------------------------- //
// Extensions
// -------------------------------------------------------------------------- //

extension Array where Element: Comparable {
    var median: Element? {
        guard !isEmpty else { return nil }
        return sorted()[count / 2]
    }
}

extension String {
    var isValidEmail: Bool { contains("@") && contains(".") }

    func camelToSnake() -> String {
        let pattern = "([A-Z])"
        let regex = try? NSRegularExpression(pattern: pattern)
        let range = NSRange(startIndex..., in: self)
        return regex?
            .stringByReplacingMatches(in: self, range: range, withTemplate: "_$1")
            .lowercased()
            .trimmingCharacters(in: CharacterSet(charactersIn: "_")) ?? self
    }
}

extension Sequence {
    func groupBy<Key: Hashable>(_ keyFn: (Element) -> Key) -> [Key: [Element]] {
        reduce(into: [:]) { $0[keyFn($1), default: []].append($1) }
    }
}

// -------------------------------------------------------------------------- //
// Actors
// -------------------------------------------------------------------------- //

actor MemoryStore<T: Entity & Sendable>: Repository {
    private var store: [UUID: T] = [:]

    func findById(_ id: UUID) async throws -> T {
        guard let entity = store[id] else { throw AppError.notFound(resource: "Entity", id: id.uuidString) }
        return entity
    }

    func list() async throws -> [T] { Array(store.values) }

    func save(_ entity: T) async throws -> T {
        store[entity.id] = entity
        return entity
    }

    func delete(_ id: UUID) async throws { store.removeValue(forKey: id) }
}

// -------------------------------------------------------------------------- //
// Async / await
// -------------------------------------------------------------------------- //

func fetchJSON<T: Decodable>(url: URL) async throws -> T {
    let (data, response) = try await URLSession.shared.data(from: url)
    guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
        throw AppError.network(statusCode: (response as? HTTPURLResponse)?.statusCode ?? 0, url: url)
    }
    return try JSONDecoder().decode(T.self, from: data)
}

func withRetry<T>(
    attempts: Int = maxRetries,
    operation: () async throws -> T
) async throws -> T {
    var lastError: Error?
    for attempt in 0..<attempts {
        do { return try await operation() }
        catch { lastError = error; try await Task.sleep(nanoseconds: UInt64(100_000_000) << attempt) }
    }
    throw lastError!
}

// -------------------------------------------------------------------------- //
// Result builders
// -------------------------------------------------------------------------- //

@resultBuilder
struct ArrayBuilder<T> {
    static func buildBlock(_ components: T...) -> [T] { components }
    static func buildOptional(_ component: [T]?) -> [T] { component ?? [] }
    static func buildEither(first: [T]) -> [T] { first }
    static func buildEither(second: [T]) -> [T] { second }
}

func makeList<T>(@ArrayBuilder<T> build: () -> [T]) -> [T] { build() }

// -------------------------------------------------------------------------- //
// Pattern matching
// -------------------------------------------------------------------------- //

func classify(_ value: Any) -> String {
    switch value {
    case let n as Int where n < 0:     return "negative int"
    case let n as Int:                 return "non-negative int \(n)"
    case let s as String where s.isEmpty: return "empty string"
    case let s as String:              return "string(\(s))"
    case let p as Point where p == .origin: return "origin"
    case let p as Point:               return "point\(p)"
    case is [Any]:                     return "array"
    default:                           return "unknown(\(type(of: value)))"
    }
}

// -------------------------------------------------------------------------- //
// KeyPaths
// -------------------------------------------------------------------------- //

func pluck<T, V>(_ keyPath: KeyPath<T, V>) -> (T) -> V {
    { $0[keyPath: keyPath] }
}

// -------------------------------------------------------------------------- //
// Fibonacci using sequence
// -------------------------------------------------------------------------- //

let fibonacci = sequence(state: (0, 1)) { state -> Int in
    let value = state.0
    state = (state.1, state.0 + state.1)
    return value
}

// -------------------------------------------------------------------------- //
// Entry point
// -------------------------------------------------------------------------- //

let user = User(name: "Alice", email: "alice@example.com")
print(user.isValid)

let fibs = Array(fibonacci.prefix(10))
print(fibs)

let p = Point(x: 3, y: 4)
print(p.distance(to: .origin))

var cfg = Config()
cfg.port = 70000  // clamped to 65535
print(cfg.baseURL)

let items = makeList { 1; 2; 3 }
print(items)

File: testdata/fixtures/src/sample.rs
//! sample.rs — comprehensive Rust syntax fixture for parser testing.
//! Covers: structs, enums, traits, generics, lifetimes, closures, iterators,
//! pattern matching, error handling, async/await, macros, trait objects,
//! smart pointers, unsafe blocks, modules, derive macros.

#![allow(dead_code, unused_variables)]

use std::{
    collections::HashMap,
    fmt,
    future::Future,
    ops::{Add, Mul},
    pin::Pin,
    sync::{Arc, Mutex},
};

// -------------------------------------------------------------------------- //
// Constants
// -------------------------------------------------------------------------- //

const MAX_RETRIES: u32 = 3;
const PI: f64 = std::f64::consts::PI;

// -------------------------------------------------------------------------- //
// Enums
// -------------------------------------------------------------------------- //

#[derive(Debug, Clone, PartialEq, Eq)]
enum Status {
    Pending,
    Running { started_at: u64 },
    Done(String),
    Failed { code: i32, reason: String },
}

impl Status {
    fn is_terminal(&self) -> bool {
        matches!(self, Status::Done(_) | Status::Failed { .. })
    }
}

impl fmt::Display for Status {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Status::Pending => write!(f, "pending"),
            Status::Running { started_at } => write!(f, "running since {started_at}"),
            Status::Done(msg) => write!(f, "done: {msg}"),
            Status::Failed { code, reason } => write!(f, "failed({code}): {reason}"),
        }
    }
}

// -------------------------------------------------------------------------- //
// Custom error type
// -------------------------------------------------------------------------- //

#[derive(Debug)]
enum AppError {
    NotFound { resource: String, id: String },
    Validation(String),
    Io(std::io::Error),
}

impl fmt::Display for AppError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            AppError::NotFound { resource, id } => {
                write!(f, "{resource} with id={id:?} not found")
            }
            AppError::Validation(msg) => write!(f, "validation error: {msg}"),
            AppError::Io(e) => write!(f, "io error: {e}"),
        }
    }
}

impl std::error::Error for AppError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        match self {
            AppError::Io(e) => Some(e),
            _ => None,
        }
    }
}

impl From<std::io::Error> for AppError {
    fn from(e: std::io::Error) -> Self {
        AppError::Io(e)
    }
}

type Result<T> = std::result::Result<T, AppError>;

// -------------------------------------------------------------------------- //
// Traits
// -------------------------------------------------------------------------- //

trait Repository<T, ID> {
    fn find_by_id(&self, id: &ID) -> Option<&T>;
    fn save(&mut self, entity: T);
    fn delete(&mut self, id: &ID) -> bool;
    fn list(&self) -> Vec<&T>;
}

trait Validate {
    fn validate(&self) -> Result<()>;
}

trait Describable: fmt::Display + fmt::Debug {}

// -------------------------------------------------------------------------- //
// Structs
// -------------------------------------------------------------------------- //

#[derive(Debug, Clone)]
struct User {
    id: String,
    name: String,
    email: String,
}

impl User {
    fn new(id: impl Into<String>, name: impl Into<String>, email: impl Into<String>) -> Self {
        User {
            id: id.into(),
            name: name.into(),
            email: email.into(),
        }
    }
}

impl Validate for User {
    fn validate(&self) -> Result<()> {
        if self.name.is_empty() {
            return Err(AppError::Validation("name must not be empty".into()));
        }
        if !self.email.contains('@') {
            return Err(AppError::Validation(
                format!("email {:?} is invalid", self.email),
            ));
        }
        Ok(())
    }
}

impl fmt::Display for User {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "User({}, {})", self.name, self.email)
    }
}

impl Describable for User {}

// -------------------------------------------------------------------------- //
// In-memory repository
// -------------------------------------------------------------------------- //

struct MemStore<T> {
    data: HashMap<String, T>,
}

impl<T> MemStore<T> {
    fn new() -> Self {
        MemStore { data: HashMap::new() }
    }
}

impl<T: Clone> Repository<T, String> for MemStore<T> {
    fn find_by_id(&self, id: &String) -> Option<&T> {
        self.data.get(id)
    }

    fn save(&mut self, entity: T) where T: Clone {
        // key extraction not available generically — example placeholder
    }

    fn delete(&mut self, id: &String) -> bool {
        self.data.remove(id).is_some()
    }

    fn list(&self) -> Vec<&T> {
        self.data.values().collect()
    }
}

// -------------------------------------------------------------------------- //
// Generics with multiple bounds & where clauses
// -------------------------------------------------------------------------- //

fn largest<T>(list: &[T]) -> &T
where
    T: PartialOrd,
{
    let mut max = &list[0];
    for item in list.iter() {
        if item > max {
            max = item;
        }
    }
    max
}

fn zip_with<A, B, C, F>(a: Vec<A>, b: Vec<B>, f: F) -> Vec<C>
where
    F: Fn(A, B) -> C,
{
    a.into_iter().zip(b).map(|(x, y)| f(x, y)).collect()
}

// -------------------------------------------------------------------------- //
// Lifetimes
// -------------------------------------------------------------------------- //

struct StrSplit<'a, 'b> {
    haystack: &'a str,
    delimiter: &'b str,
}

impl<'a, 'b> Iterator for StrSplit<'a, 'b> {
    type Item = &'a str;

    fn next(&mut self) -> Option<Self::Item> {
        if let Some(pos) = self.haystack.find(self.delimiter) {
            let part = &self.haystack[..pos];
            self.haystack = &self.haystack[pos + self.delimiter.len()..];
            Some(part)
        } else if self.haystack.is_empty() {
            None
        } else {
            let rest = self.haystack;
            self.haystack = "";
            Some(rest)
        }
    }
}

fn longest<'a>(x: &'a str, y: &'a str) -> &'a str {
    if x.len() >= y.len() { x } else { y }
}

// -------------------------------------------------------------------------- //
// Closures & iterators
// -------------------------------------------------------------------------- //

fn fibonacci() -> impl Iterator<Item = u64> {
    let mut state = (0u64, 1u64);
    std::iter::from_fn(move || {
        let next = state.0;
        state = (state.1, state.0.saturating_add(state.1));
        Some(next)
    })
}

fn pipeline_example(data: Vec<i32>) -> Vec<i32> {
    data.into_iter()
        .filter(|&x| x > 0)
        .map(|x| x * 2)
        .take_while(|&x| x < 100)
        .collect()
}

// -------------------------------------------------------------------------- //
// Smart pointers
// -------------------------------------------------------------------------- //

type SharedCounter = Arc<Mutex<u64>>;

fn make_counter() -> SharedCounter {
    Arc::new(Mutex::new(0))
}

fn increment(counter: &SharedCounter) {
    let mut guard = counter.lock().unwrap();
    *guard += 1;
}

// -------------------------------------------------------------------------- //
// Trait objects (dynamic dispatch)
// -------------------------------------------------------------------------- //

fn print_all(items: &[Box<dyn fmt::Display>]) {
    for item in items {
        println!("{item}");
    }
}

fn make_adder(x: i32) -> Box<dyn Fn(i32) -> i32> {
    Box::new(move |y| x + y)
}

// -------------------------------------------------------------------------- //
// Async / await
// -------------------------------------------------------------------------- //

async fn fetch(url: &str) -> Result<Vec<u8>> {
    // Simulated async I/O
    Ok(url.as_bytes().to_vec())
}

async fn fetch_all(urls: Vec<&str>) -> Vec<Result<Vec<u8>>> {
    let mut results = Vec::with_capacity(urls.len());
    for url in urls {
        results.push(fetch(url).await);
    }
    results
}

// -------------------------------------------------------------------------- //
// Operator overloading
// -------------------------------------------------------------------------- //

#[derive(Debug, Clone, Copy, PartialEq)]
struct Vec2 {
    x: f64,
    y: f64,
}

impl Vec2 {
    fn new(x: f64, y: f64) -> Self { Vec2 { x, y } }
    fn length(self) -> f64 { (self.x * self.x + self.y * self.y).sqrt() }
    fn dot(self, other: Vec2) -> f64 { self.x * other.x + self.y * other.y }
}

impl Add for Vec2 {
    type Output = Vec2;
    fn add(self, rhs: Vec2) -> Vec2 { Vec2::new(self.x + rhs.x, self.y + rhs.y) }
}

impl Mul<f64> for Vec2 {
    type Output = Vec2;
    fn mul(self, rhs: f64) -> Vec2 { Vec2::new(self.x * rhs, self.y * rhs) }
}

impl fmt::Display for Vec2 {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "({:.2}, {:.2})", self.x, self.y)
    }
}

// -------------------------------------------------------------------------- //
// Macros
// -------------------------------------------------------------------------- //

macro_rules! hash_map {
    ($($key:expr => $val:expr),* $(,)?) => {{
        let mut map = HashMap::new();
        $(map.insert($key, $val);)*
        map
    }};
}

macro_rules! assert_ok {
    ($expr:expr) => {
        match $expr {
            Ok(v) => v,
            Err(e) => panic!("expected Ok, got Err({e:?})"),
        }
    };
}

// -------------------------------------------------------------------------- //
// Unsafe
// -------------------------------------------------------------------------- //

fn raw_pointer_demo() {
    let mut val = 42i32;
    let raw = &mut val as *mut i32;
    unsafe {
        *raw += 1;
        assert_eq!(*raw, 43);
    }
}

// -------------------------------------------------------------------------- //
// Pattern matching advanced
// -------------------------------------------------------------------------- //

fn classify(n: i64) -> &'static str {
    match n {
        i64::MIN..=-1 => "negative",
        0 => "zero",
        1..=9 => "single digit",
        10..=99 => "double digit",
        _ if n % 2 == 0 => "large even",
        _ => "large odd",
    }
}

// -------------------------------------------------------------------------- //
// Entry point
// -------------------------------------------------------------------------- //

fn main() {
    let user = User::new("u1", "Alice", "alice@example.com");
    println!("{user}");
    assert_ok!(user.validate());

    let fibs: Vec<u64> = fibonacci().take(10).collect();
    println!("{fibs:?}");

    let v1 = Vec2::new(1.0, 0.0);
    let v2 = Vec2::new(0.0, 1.0);
    println!("v1 + v2 = {}", v1 + v2);

    let counter = make_counter();
    increment(&counter);
    println!("counter = {}", counter.lock().unwrap());

    let m = hash_map!["one" => 1, "two" => 2, "three" => 3];
    println!("{m:?}");
}
